pipeline/nissl_to_ccfv3.py

```diff
@@ -0,0 +1,223 @@
+#!/usr/bin/env python
+# Copyright 2021, Blue Brain Project, EPFL
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Script that maps Nissl to CCFv3."""
+from __future__ import annotations
+
+import argparse
+import logging
+import pathlib
+import sys
+from pathlib import Path
+
+import numpy as np
+
+from atlannot import load_volume
+from atlannot.ants import register, transform
+
+logger = logging.getLogger("nissl-to-ccfv3")
+
+
+def parse_args():
+    """Parse arguments."""
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--nissl-path",
+        type=Path,
+        help="""\
+        Path to Nissl Volume.
+        """,
+    )
+    parser.add_argument(
+        "--ccfv2-path",
+        type=Path,
+        help="""\
+        Path to CCFv2 annotation volume.
+        """,
+    )
+    parser.add_argument(
+        "--ccfv3-path",
+        type=Path,
+        help="""\
+        Path to CCFv3 annotation volume.
+        """,
+    )
+    parser.add_argument(
+        "--output-dir",
+        type=Path,
+        help="""\
+        Path to output directory to save results.
+        """,
+    )
+    return parser.parse_args()
+
+
+def check_and_load(path: pathlib.Path | str) -> np.array:
+    """Load volume if path exists.
+
+    Parameters
+    ----------
+    path
+        File path.
+
+    Returns
+    -------
+    volume : np.array
+        Loaded volume.
+    """
+    path = Path(path)
+    if not path.exists():
+        logger.error(f"The specified path {path} does not exist.")
+        return 1
+    volume = load_volume(path, normalize=False)
+    return volume.astype(np.float32)
+
+
+def slice_registration(
+    fixed: np.array, moving: np.array, nissl_slice: np.array | None = None
+) -> tuple[np.array, np.array | None]:
+    """Compute registration transform between a couple of slices.
+
+    Parameters
+    ----------
+    fixed
+        Fixed slice of shape (h, w) of type annotation.
+    moving
+        Moving slice of shape (h, w) of type annotation.
+    nissl_slice
+        If needed, intensity slice of shape (h, w)  in the same coordinate
+        system as moving.
+
+    Returns
+    -------
+    warped : np.array
+        Warped slice of type annotation.
+    nissl_slice : np.array | None
+        If a nissl slice is specified, the given slice is warped with the
+        same transformation.
+    """
+    nii_data = register(fixed, moving, is_atlas=True)
```

Comment:
I got the following ?warning? locally when running the whole script. It did not kill the script.

```
Exception Object caught: 

itk::ExceptionObject (0x1022e7bf0)
Location: "unknown" 
File: /private/var/folders/sd/t2l8818d3qb3chf72yjvp36n5q8_d1/T/pip-install-9su5jrb4/antspyx_154ddc815c3a4e6bb7a5030d63b97590/itksource/Modules/Filtering/ImageStatistics/include/itkImageMomentsCalculator.hxx
Line: 124
Description: ITK ERROR: ImageMomentsCalculator(0x600001f8c960): Compute(): Total Mass of the image was zero. Aborting here to prevent division by zero later on.


Exception Object caught: 

itk::ExceptionObject (0x1022e7bf0)
Location: "unknown" 
File: /private/var/folders/sd/t2l8818d3qb3chf72yjvp36n5q8_d1/T/pip-install-9su5jrb4/antspyx_154ddc815c3a4e6bb7a5030d63b97590/itksource/Modules/Filtering/ImageStatistics/include/itkImageMomentsCalculator.hxx
Line: 124
Description: ITK ERROR: ImageMomentsCalculator(0x600001f8c3c0): Compute(): Total Mass of the image was zero. Aborting here to prevent division by zero later on.

```
Have you seen it too?

Reply:
It is because some Nissl slices are really bad quality. I don't think we can really do something about this.